internal/consensus/state.go

```diff
@@ -1373,7 +1374,10 @@ func (cs *State) timelyProposalMargins() (time.Duration, time.Duration) {
 
 	// cs.ProposalReceiveTime - cs.Proposal.Timestamp >= -1 * Precision
 	// cs.ProposalReceiveTime - cs.Proposal.Timestamp <= MessageDelay + Precision
-	return -sp.Precision, sp.MessageDelay + sp.Precision
+	if sp.MessageDelay+sp.Precision < 0 { // overflow check: need for some architectures
+		return -sp.Precision, time.Duration(math.MaxInt64)
+	}
+	return -sp.Precision, time.Duration(float64(sp.MessageDelay) + float64(sp.Precision))
```

Comment:
Why are we converting to float64 here and then back to a duration?

It seems like this should be sufficient.

```go
	if sp.MessageDelay < math.MaxInt64-sp.Precision {
		return -sp.Precision, time.Duration(math.MaxInt64)
	}
	return -sp.Precision, sp.MessageDelay + sp.Precision
```

Reply:
https://go.dev/play/p/D36e4asgpur ?

Reply:
```
sum := adaptedSP.MessageDelay + adaptedSP.Precision // overflows
sum := time.Duration(adaptedSP.MessageDelay + adaptedSP.Precision) // overflows
sum := time.Duration(float64(adaptedSP.MessageDelay) + float64(adaptedSP.Precision)) // does not overflow
```

I agree that we need an additional wrapping in the presence of `if` condition. BUT it doesn't hurt to have it, just in case...

Reply:
Sorry, I have the sense wrong on the inequality.

Should be,
```go
	if sp.MessageDelay > math.MaxInt64-sp.Precision {
		return -sp.Precision, time.Duration(math.MaxInt64)
	}
	return -sp.Precision, sp.MessageDelay + sp.Precision
```

Reply:
So, this is a workaround, not a solution.

The real solution is to add some "safe add duration" method, say, to `types/time` package (typically imported as `cmttime`?

Reply:
Regarding the conversion to `float64`, the point is that in the `arm64` architecture, this does the trick. Namely, overflows are detected and handled nicely.

The problem is that this is architecture-dependent. In the case of the `linux/amd64` this does not work, for some unknown reason. This is something to report to Golang.

Reply:
> `if sp.MessageDelay > math.MaxInt64-sp.Precision`

This is more precise, indeed, but what in the case that `MessageDelay` has overflowed? It would be negative, this if of course will not be triggered, and we are returning a negative value.